### PR TITLE
Feature: Add a new Davidson iteration method called subspace davidson for pw basis

### LINKED ISSUE #3902: The current Davidson iteration method for pw basis has low computation efficiency.
### Details

We need a new Davidson method for pw basis.

### Task list for Issue attackers (only for developers)

- [ ] Reproduce the performance issue on a similar system or environment.
- [ ] Identify the specific section of the code causing the performance issue.
- [ ] Investigate the issue and determine the root cause.
- [ ] Research best practices and potential solutions for the identified performance issue.
- [ ] Implement the chosen solution to address the performance issue.
- [ ] Test the implemented solution to ensure it improves performance without introducing new issues.
- [ ] Optimize the solution if necessary, considering trade-offs between performance and other factors (e.g., code complexity, readability, maintainability).
- [ ] Review and incorporate any relevant feedback from users or developers.
- [ ] Merge the improved solution into the main codebase and notify the issue reporter.

### PR DESCRIPTION
**Pull request by haozhihan**
### Reminder
- [x] Have you linked an issue with this pull request?
- [ ] Have you added adequate unit tests and/or case tests for your pull request?
- [ ] Have you noticed possible changes of behavior below or in the linked issue?
- [ ] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Details
- At present, the method supports CPU single core and CPU multi-core parallel. In the near future, it will support for CUDA-GPU, ROCm GPU, and ROCm DCU.

### Linked Issue
Close #3902 

### Unit Tests and/or Case Tests for my changes
- There are currently no Tests available. But I will add some Tests as soon as possible in the future.

### What's changed?
- I have added a new `ks_solver` parameter called `subspace_dav` to control calls to this new iteration method.
- The main code for this method is concentrated in the `diagon_subspacedav.cpp` and `diagon_subspacedav.h` files.

### Any changes of core modules? (ignore if not applicable)
- I modified a small amount of code in `hsolver_pw.cpp`

### CODE REVIEW

File: source/module_io/input.cpp [hunk 1]
```diff
@@ -3001,7 +3001,7 @@ void Input::Default_2(void) // jiyy add 2019-08-04
                 diago_proc = GlobalV::NPROC;
             }
         }
-        else if (ks_solver == "dav")
+        else if (ks_solver == "dav" || ks_solver == "subspace_dav")
```
Comment (mohanchen) [inline source/module_io/input.cpp]: I suggest we can change the method from "subspace_dav" to "dav_subspace"

File: source/module_io/input.cpp [hunk 2]
```diff
@@ -3950,7 +3950,7 @@ void Input::Check(void)
         {
             ModuleBase::WARNING_QUIT("Input", "lapack can not be used with plane wave basis.");
         }
-        else if (ks_solver != "default" && ks_solver != "cg" && ks_solver != "dav" && ks_solver != "bpcg")
+        else if (ks_solver != "default" && ks_solver != "cg" && ks_solver != "dav" && ks_solver != "subspace_dav" && ks_solver != "bpcg")
```
Comment (mohanchen) [inline source/module_io/input.cpp]: change to several lines, each line for each "&&" option

File: source/module_hsolver/diago_subspacedav.cpp [hunk 3]
```diff
@@ -0,0 +1,976 @@
...
+                       this->dim,
+                       nbase,
+                       this->notconv,
+                       basis,
+                       this->hphi,
+                       this->vcc,
+                       unconv.data(),
+                       this->eigenvalue_in_dav);
+
+        this->cal_elem(this->dim, nbase, this->notconv, basis, this->hphi, this->hcc, this->scc, false);
```
Comment (mohanchen) [inline source/module_hsolver/diago_subspacedav.cpp]: change to one line for each input variable

File: source/module_hsolver/diago_subspacedav.cpp [hunk 4]
```diff
@@ -0,0 +1,976 @@
...
+                unconv[this->notconv] = m;
+                this->notconv++;
+            }
+
+            eigenvalue_in_hsolver[m] = this->eigenvalue_in_dav[m];
+        }
+
+        ModuleBase::timer::tick("Diago_SubspaceDav", "check_update");
+
+        if (this->notconv == 0 || (nbase + this->notconv + 1 > this->nbase_x)
```
Comment (mohanchen) [inline source/module_hsolver/diago_subspacedav.cpp]: change to one line for each option

File: source/module_hsolver/diago_subspacedav.cpp [hunk 5]
```diff
@@ -0,0 +1,976 @@
...
+                                       const int* unconv,
+                                       Real* eigenvalue_in_dav)
+{
+    ModuleBase::timer::tick("Diago_SubspaceDav", "cal_grad");
+
+    for (size_t i = 0; i < notconv; i++)
+    {
+        if (unconv[i] != i)
+        {
+            syncmem_complex_op()(this->ctx, this->ctx, vcc + i * this->nbase_x, vcc + unconv[i] * this->nbase_x, nbase);
```
Comment (mohanchen) [inline source/module_hsolver/diago_subspacedav.cpp]: this line is too long

File: source/module_hsolver/diago_subspacedav.cpp [hunk 6]
```diff
@@ -0,0 +1,976 @@
...
+    // "precondition!!!"
+    std::vector<Real> pre(this->dim, 0.0);
+    for (int m = 0; m < notconv; m++)
+    {
+        for (size_t i = 0; i < this->dim; i++)
+        {
+            double x = this->precondition[i] - this->eigenvalue_in_dav[m];
+            pre[i] = 0.5 * (1.0 + x + sqrt(1 + (x - 1.0) * (x - 1.0)));
+        }
+        vector_div_vector_op<T, Device>()(this->ctx, this->dim, &basis(nbase + m, 0), &basis(nbase + m, 0), pre.data());
```
Comment (mohanchen) [inline source/module_hsolver/diago_subspacedav.cpp]: this line is too long

File: source/module_hsolver/diago_subspacedav.cpp [hunk 7]
```diff
@@ -0,0 +1,976 @@
...
+
+    nbase += notconv;
+    int nb1 = nbase - notconv;
+    // reset:
+    if (init)
+    {
+        for (size_t i = 0; i < nbase; i++)
+        {
+
+            hcc[i * this->nbase_x + i] = set_real_tocomplex(hcc[i * this->nbase_x + i]);
```
Comment (mohanchen) [inline source/module_hsolver/diago_subspacedav.cpp]: this 'i * this->nbase_x + i' can be calculated only once and use a variable to store the result

File: source/module_hsolver/diago_subspacedav.cpp [hunk 8]
```diff
@@ -0,0 +1,976 @@
...
+            {
+                hcc[i * this->nbase_x + j] = cs.zero;
+                scc[i * this->nbase_x + j] = cs.zero;
+                hcc[j * this->nbase_x + i] = cs.zero;
+                scc[j * this->nbase_x + i] = cs.zero;
+            }
+        }
+    }
+
+    // if (!init &&  (GlobalV::RANK_IN_POOL == 0))
```
Comment (mohanchen) [inline source/module_hsolver/diago_subspacedav.cpp]: is these commented out codes are for tests, write down the explanations

File: source/module_hsolver/diago_subspacedav.cpp [hunk 9]
```diff
@@ -0,0 +1,976 @@
...
+                    s_diag[i][j] = scc[i * this->nbase_x + j];
+                }
+            }
+
+            if (this->device == psi::GpuDevice)
+            {
+#if defined(__CUDA) || defined(__ROCM)
+                Real* eigenvalue_gpu = nullptr;
+                resmem_real_op()(this->ctx, eigenvalue_gpu, this->nbase_x);
+                syncmem_var_h2d_op()(this->ctx, this->cpu_ctx, eigenvalue_gpu, this->eigenvalue_in_dav, this->nbase_x);
```
Comment (mohanchen) [inline source/module_hsolver/diago_subspacedav.cpp]: this line is too long, one variable for each line

**Comment by WHUweiqingzhou** [comment]
Should this PR also close #3447?

**Comment by haozhihan** [comment]
> Should this PR also close #3447?

I think we should wait for the code to merge and test the cases mentioned in the issue #3447 again, comparing it with the previous efficiency. Then we can decide whether to close this issue.

File: source/module_hsolver/hsolver_pw.cpp [hunk 10]
```diff
@@ -244,10 +287,10 @@ void HSolverPW<T, Device>::solve(hamilt::Hamilt<T, Device>* pHamilt,
...
-            ModuleBase::Vector3<double> *_gk = new ModuleBase::Vector3<double>[npw];
-            for (int ig = 0;ig < npw; ig++)
+            ModuleBase::Vector3<double>* _gk = new ModuleBase::Vector3<double>[npw];
+            for (int ig = 0; ig < npw; ig++)
             {
-                _gk[ig] = this->wfc_basis->getgpluskcar(ik,ig);
+                _gk[ig] = this->wfc_basis->getgpluskcar(ik, ig);
             }
 
             double* kpt;
```
Comment (mohanchen) [inline source/module_hsolver/hsolver_pw.cpp:296]: kpt = nullptr;

**Review by mohanchen — COMMENTED** [review]
LGTM, make sure you have used clang format

**Comment by haozhihan** [comment]
> LGTM, make sure you have used clang format

Yes, I have.

**Review by mohanchen — APPROVED** [review]
LGTM

**Pull request merged**